ssl/ech/ech_ssl_apis.c

```diff
@@ -13,70 +13,311 @@
 
 int SSL_CTX_set1_echstore(SSL_CTX *ctx, OSSL_ECHSTORE *es)
 {
-    return 0;
+    if (ctx == NULL || es == NULL) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    OSSL_ECHSTORE_free(ctx->ext.ech.es);
+    ctx->ext.ech.es = NULL;
+    if (ossl_echstore_dup(&ctx->ext.ech.es, es) != 1) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    return 1;
 }
 
-int SSL_set1_echstore(SSL *s, OSSL_ECHSTORE *es)
+int SSL_set1_echstore(SSL *ssl, OSSL_ECHSTORE *es)
 {
-    return 0;
+    SSL_CONNECTION *s;
+
+    s = SSL_CONNECTION_FROM_SSL_ONLY(ssl);
+    if (s == NULL || es == NULL)
+        return 0;
+    OSSL_ECHSTORE_free(s->ext.ech.es);
+    s->ext.ech.es = NULL;
+    if (ossl_echstore_dup(&s->ext.ech.es, es) != 1) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    /*
+     * Here, and below, if the application calls an API that implies it
+     * wants to try ECH, then we set attempted to 1
+     */
+    s->ext.ech.attempted = 1;
+    return 1;
 }
 
 OSSL_ECHSTORE *SSL_CTX_get1_echstore(const SSL_CTX *ctx)
 {
-    return NULL;
+    OSSL_ECHSTORE *dup = NULL;
+
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return NULL;
+    }
+    if (ossl_echstore_dup(&dup, ctx->ext.ech.es) != 1) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
```

Comment:
It doesn't seem like an "internal error" if the application has called `SSL_CTX_get1_echstore` without first calling `SSL_CTX_set1_echstore` (i.e. `ctx->ext.ech.es` is NULL).

Reply:
sure, so just return NULL and don't raise any error? doing that...